### PR TITLE
fix file specific cache control and expires opts (ports #958)

### PR DESCRIPTION
**Pull request by svenfuchs**
Can be tested using:

```yaml
deploy:
  provider: s3
  edge:
    branch: sf-s3-file-specific-opts
⋮
```

### CODE REVIEW

File: lib/dpl/providers/s3.rb [hunk 1]
```diff
@@ -30,8 +30,8 @@ class S3 < Provider
       opt '--server_side_encryption', 'Use S3 Server Side Encryption (SSE-AES256)'
       opt '--local_dir DIR', 'Local directory to upload from', default: '.', example: '~/travis/build (absolute path) or ./build (relative path)'
       opt '--detect_encoding', 'HTTP header Content-Encoding for files compressed with gzip and compress utilities'
-      opt '--cache_control STR', 'HTTP header Cache-Control to suggest that the browser cache the file', default: 'no-cache', enum: ['no-cache', 'no-store', /max-age=\d+/, /s-maxage=\d+/, 'no-transform', 'public', 'private']
-      opt '--expires DATE', 'Date and time that the cached object expires', format: /^\d{4}-\d{2}-\d{2} \d{2}:\d{2}:\d{2} .+$/
+      opt '--cache_control STR', 'HTTP header Cache-Control to suggest that the browser cache the file', type: :array, default: 'no-cache', enum: [/^no-cache.*/, /^no-store.*/, /^max-age=\d+.*/, /^s-maxage=\d+.*/, /^no-transform/, /^public/, /^private/], note: 'accepts mapping values to globs', eg: 'publich: *.css,*.js'
```
Comment (viniciuskneves) [inline lib/dpl/providers/s3.rb]: Typo here `public` instead of `publich`

**Comment by viniciuskneves** [comment]
@svenfuchs will it keep this same structure? https://github.com/travis-ci/dpl/pull/958#issuecomment-467197692

As I understood yes, just want to make sure. Also, is it documented somewhere?

File: lib/dpl/providers/s3.rb (hunk 1 above)
Comment (svenfuchs) [reply lib/dpl/providers/s3.rb]: good catch! fixed.

**Comment by svenfuchs** [comment]
@viniciuskneves not precisely, but almost. that's because the nested array in this config would require changes to travis-build, and commit to a semantical meaning of nested arrays here that i'd rather avoid making:

```
 - provider: s3
    cache_control:
      - "max-age=3000":
        - "*.css"
        - "*.png"
```

instead you can either use either of these:

```
    cache_control:
      - "max-age=3000: *.css, *png"
      - "max-age=3000: *.css,*png"
      - "max-age=3000: *.{css,png}"
```

or list them separately:

```
    cache_control:
      - "max-age=3000: *.css"
      - "max-age=3000: *.png"
```

do you also think this is functionally equivalent?

**Comment by svenfuchs** [comment]
ooops, missed your other question, sorry. 

it's not yet documented anywhere besides what's on the option definition. we should probably talk about this extensively in our end user documentation. not sure the readme needs a lot more, but if it does we could use the provider's `description` for that.

**Comment by chriselsen** [comment]
Awesome to finally see some movement on this item! Thank you. 

Works perfect for me with the following config: 

```
cache_control: 
      - "public, max-age=31536000: *.jpg, *.gif, *.png, *.webp, *.css, *.js, *.ico"
expires:
      - "\"`date -u +\"%Y-%m-%d %H:%M:%S -0000\" --date \"next Year\"`\": *.jpg, *.gif, *.png, *.webp, *.css, *.js, *.ico"
```

Only small glitch: If I set a cache_control header twice for the same file types, I would have expected to only take the second line. But instead only the first line is applied. E.g.
```
cache_control: 
      - "public: *.jpg, *.gif, *.png, *.webp, *.css, *.js, *.ico"
      - "max-age=31536000: *.jpg, *.gif, *.png, *.webp, *.css, *.js, *.ico"
```

Great work!

**Comment by svenfuchs** [comment]
@chriselsen thanks for the feedback, great to know this works for your usecase!

hmmm, yes, interesting. it picks the first match that it finds, i somehow thought that would be an intuitive choice.

do you think documenting this behavior would be good enough? or do you feel strongly about picking the last match?

**Comment by chriselsen** [comment]
> hmmm, yes, interesting. it picks the first match that it finds, i somehow thought that would be an intuitive choice.
> 
> do you think documenting this behavior would be good enough? or do you feel strongly about picking the last match?

I was thinking along the lines of a variable being overwritten if it is assigned again. Although documenting the current behavior could be enough. 
Let me know how I can help with updating the documentation. 

Really love to see this capability shipped!

**Comment by svenfuchs** [comment]
ok, great, thanks for the feedback, @chriselsen @viniciuskneves! i'll merge this then.

we might not end up mentioning this in the dpl v2 developer preview release blog post that we'll hopefully do this week because of the missing docs and such. i have put those on my todo list though, and look into adding docs asap.

**Comment by viniciuskneves** [comment]
Thank you @svenfuchs 

I didn't have time to give it a try yet but looks really good from what I've seen.

Could you also @ in the docs? I think it is important to have this behavior really well documented.

**Comment by chriselsen** [comment]
I had another closer look and it actually doesn't seem to be working:

You can see above the config that I'm using. 
Looking at the logs for a deploy of my previous patch, I can see that the cache_control and expires header are set correctly on a matching object:  
```
uploading "assets/images/og-image.jpg" with {:content_type=>"image/jpeg", :cache_control=>"public, max-age=31536000", :acl=>"public-read", :expires=>"2020-08-16 18:30:13 -0000"}
```
At the same time for a non-matching object these headers are not set:
```
uploading "tags/index.html" with {:content_type=>"text/html; charset=utf-8", :cache_control=>nil, :acl=>"public-read", :expires=>nil}
```
With the new DPL2 version the cache_control and expires header are set to the entire line specified in the .travis.yml config - thus including the file-extensions - on all objects:
```
uploading "assets/images/og-image.jpg" with {:content_type=>"image/jpeg", :cache_control=>"\"public, max-age=31536000\": *.jpg, *.gif, *.png, *.webp, *.css, *.js, *.ico", :acl=>"public-read", :expires=>"\"2020-08-20 17:37:53 -0000\": *.jpg, *.gif, *.png, *.webp, *.css, *.js, *.ico"}
uploading "tags/index.html" with {:content_type=>"text/html; charset=utf-8", :cache_control=>"\"public, max-age=31536000\": *.jpg, *.gif, *.png, *.webp, *.css, *.js, *.ico", :acl=>"public-read", :expires=>"\"2020-08-20 17:37:53 -0000\": *.jpg, *.gif, *.png, *.webp, *.css, *.js, *.ico"}
```

The same happens even in the case of a super simple configuration of:
```
cache_control:
      - "max-age=3000: *.png"
```

**Comment by svenfuchs** [comment]
wow, ok. that's strange. i would have thought that's something the unit tests should have caught. 

i'll take a look. thanks for the report @chriselsen!

**Comment by svenfuchs** [comment]
@chriselsen i'm failing to reproduce this, and the unit tests in fact seem to suggest it works ... so i most probably misunderstand some detail about your build/config.

here's a build that has uploaded a file `dpl_test` with the following config https://travis-ci.com/travis-ci/dpl/builds/123942476#L183

```
deploy:
  - provider: s3
    access_key_id: ...
    secret_access_key: ...
    bucket: travis-dpl-test
    local_dir: ./tmp/dpl_test
    cache_control: 'max-age=3000: *'
```

Do you have a url for your build log and `.travis.yml`? Or if it's private, could you paste or gist the relevant parts?
The build log, but also `curl` confirms the header has been set as expected:

```
$ curl -v https://s3.amazonaws.com/travis-dpl-test/dpl_test
*   Trying 52.216.204.213...
* TCP_NODELAY set
* Connected to s3.amazonaws.com (52.216.204.213) port 443 (#0)
* TLS 1.2 connection using TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256
* Server certificate: s3.amazonaws.com
* Server certificate: DigiCert Baltimore CA-2 G2
* Server certificate: Baltimore CyberTrust Root
> GET /travis-dpl-test/dpl_test HTTP/1.1
> Host: s3.amazonaws.com
> User-Agent: curl/7.61.0
> Accept: */*
>
< HTTP/1.1 200 OK
< x-amz-id-2: ND17hcJ2vQmQw7VYPcCyx+nSykm52sQXR4hGNH5HFM+8QCbKqsyy5N0f/qq0F0YNCeXQoiYwWtc=
< x-amz-request-id: CA4E915A71074C09
< Date: Tue, 20 Aug 2019 18:36:18 GMT
< Last-Modified: Tue, 20 Aug 2019 18:33:48 GMT
< ETag: "c11168d120515612af61b787ad3e5238"
< Cache-Control: max-age=3000
< Accept-Ranges: bytes
< Content-Type: application/octet-stream
< Content-Length: 32
< Server: AmazonS3
<
* Connection #0 to host s3.amazonaws.com left intact
0db167fbeec6717b96e84f31e3f6c9f0% 
```

**Comment by chriselsen** [comment]
@svenfuchs Found the issue!

Looks like this config file works:
```
cache_control: "\"public, max-age=31536000\": *.jpg, *.gif, *.png, *.webp, *.css, *.js, *.ico"
    expires: "\"`date -u +\"%Y-%m-%d %H:%M:%S -0000\" --date \"next Year\"`\": *.jpg, *.gif, *.png, *.webp, *.css, *.js, *.ico"
```
See https://travis-ci.com/edge-cloud/www.edge-cloud.net/builds/123548904

While this config file does not work:
```
cache_control: 
      - "public, max-age=31536000: *.jpg, *.gif, *.png, *.webp, *.css, *.js, *.ico"
    expires: 
      - "`date -u +\"%Y-%m-%d %H:%M:%S -0000\" --date \"next Year\"`: *.jpg, *.gif, *.png, *.webp, *.css, *.js, *.ico"
```
See https://travis-ci.com/edge-cloud/www.edge-cloud.net/builds/123947367

Doh, I just re-read my own comments on https://github.com/travis-ci/dpl/issues/486 from February and it makes sense that the second approach doesn't work as travis-yml isn't setup to correctly pass the structure on to dpl. 
Guess that shows again the importance of docs. 
Want me to submit a pull request for https://github.com/travis-ci/dpl#aws-s3 with some examples?

**Comment by svenfuchs** [comment]
@chriselsen super, glad you've figured it out!

the README is autogenerated, but yes, feel free to add some examples to the option definition.

i think it might make even more sense to start with a PR for https://github.com/travis-ci/docs-travis-ci-com/blob/master/user/deployment/s3.md though, as that's what most people are looking at.

in any case, thanks a lot for the great help!

**Comment by chriselsen** [comment]
Here are the pull requests for the documentation updates:
https://github.com/travis-ci/docs-travis-ci-com/pull/2490
https://github.com/travis-ci/dpl/pull/1075

**Comment by toxic-johann** [comment]
Hi @chriselsen @svenfuchs  , I add cache-control in my `travis.yml` as your example. But it doesn't work.

My config looks like below:

```
deploy:
  provider: s3
  skip_cleanup: true
  acl: private
  local_dir: build
  cache_control: "\"public, max-age=31536000\": *.jpg, *.gif, *.png, *.webp, *.css, *.js, *.ico"
  expires: "\"`date -u +\"%Y-%m-%d %H:%M:%S -0000\" --date \"next Year\"`\": *.jpg, *.gif, *.png, *.webp, *.css, *.js, *.ico"
  on:
    all_branches: true
    condition: $TUBI_DEPLOYMENT = true
```

But it finally include all objects in cache-control value.

image

I have try all the example you talk in this issue, other issues and the document, but all of them don't works. Any idea how to fix this?

**Comment by chriselsen** [comment]
DPL v2 hasn't been [released](https://github.com/travis-ci/dpl/releases) yet. In the meantime you need to manually specify in your .travis.yml that you want to use the latest version of DPL for this to work:
```
deploy:
  - provider: s3
    edge:
      source: travis-ci/dpl
      branch: master
```

**Pull request merged**